gh-93735: Split Docs CI to speed-up the build

AA-Turner opened the pull request:
#93735

A

AA-Turner committed:
Speed-up docs CI

hugovk commented on .github/workflows/doc.yml:
```diff
@@ -32,6 +33,37 @@ jobs:
     - name: Register Sphinx problem matcher
       run: echo "::add-matcher::.github/problem-matchers/sphinx.json"
+    - name: 'Setup Python'
+      uses: actions/setup-python@v3
```
We can bump to the new v4
```suggestion
      uses: actions/setup-python@v4
```

AA-Turner replied:
I entirely missed v4 was even released!

hugovk replied:
Oh, and it needs a `python-version`, can be `"3.x"`:

https://github.com/actions/setup-python/releases/tag/v4.0.0

AA-Turner replied:
Jinx https://github.com/python/cpython/pull/93736#issuecomment-1153009014 :P

hugovk commented on .github/workflows/doc.yml:
```diff
@@ -32,6 +33,37 @@ jobs:
+    name: 'Doctest'
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
```
```suggestion
    - uses: actions/checkout@v4
```

ezio-melotti replied:
This was marked as resolved, but the diff still shows `@v3`

AA-Turner replied:
There is no checkout v4 action, hence marking as resolved.

A

hugovk commented on .github/workflows/doc.yml:
```diff
@@ -32,6 +33,37 @@ jobs:
     - uses: actions/checkout@v3
     - name: Register Sphinx problem matcher
       run: echo "::add-matcher::.github/problem-matchers/sphinx.json"
+    - name: 'Setup Python'
```
"Setup" is a noun, this is with verb:
```suggestion
    - name: 'Set up Python'
```

hugovk reviewed (commented):
A nice speed improvement!

Looking at an [earlier build](https://github.com/python/cpython/runs/6845334202?check_suite_focus=true) (~15 mins):

<img width="1113" alt="image" src="https://user-images.githubusercontent.com/1324225/173206322-6d466260-65f8-4e53-be9d-eaf98e9596ff.png">

This has parallel docs (~4 min) and doctest (~7 min):

<img width="1127" alt="image" src="https://user-images.githubusercontent.com/1324225/173206338-0378c122-d86b-42f2-bf53-85dd8e5d1544.png">

and

<img width="1126" alt="image" src="https://user-images.githubusercontent.com/1324225/173206353-e1d630ba-a3ab-4e39-91e4-7646437228e1.png">

So even if they ran in serial (they won't), it's still quicker. In part it's because "Build HTML documentation" only takes 2 min instead of 6 min (or sometimes up to 15 min)!

AA-Turner committed:
Bump actions; grammar; both per Hugo

AA-Turner committed:
Explicit python version

ezio-melotti commented on .github/workflows/doc.yml:
```diff
@@ -32,6 +33,38 @@ jobs:
+    - name: 'Check documentation'
+      run: make -C Doc/ SPHINXOPTS="-q -W --keep-going" check
+    - name: 'Build HTML documentation'
+      run: make -C Doc/ SPHINXOPTS="-q -W --keep-going" html
```
The `-W` since to be already included.  The output command shows:
```
PATH=./venv/bin:$PATH sphinx-build -b html -d build/doctrees  -j auto -q -W --keep-going -W . build/html 
```
(Not your fault since you were just copy-pasting, but maybe it can be removed here and above?)

AA-Turner replied:
`-W` comes from SPHINXERRORHANDLING, I moved `-W` and `--keep-going` to SPHINXERRORHANDLING (If better to just put `-q --keep-going` in SPHINXOPTS I'll do so, I don't know if it would be possible for the `-W` to be lost somehow in other make configuration)

A

ezio-melotti commented on .github/workflows/doc.yml:
```diff
@@ -32,6 +33,38 @@ jobs:
+    - name: 'Install build dependencies'
+      run: make -C Doc/ venv
+    - name: 'Check documentation'
+      run: make -C Doc/ SPHINXOPTS="-q -W --keep-going" check
```
This is now using `sphinx-lint`, so it doesn't need the `SPHINXOPTS`.

ezio-melotti commented on .github/workflows/doc.yml:
```diff
@@ -32,6 +33,38 @@ jobs:
+    steps:
+    - uses: actions/checkout@v3
+    - name: Register Sphinx problem matcher
+      run: echo "::add-matcher::.github/problem-matchers/sphinx.json"
```
How does this lone `echo` work?

AA-Turner replied:
Again, copy-pasting! It is actions syntax for the problem matchers, the 'echo' is caught by GHA's log interpreter.

A

hugovk replied:
Docs: https://github.com/actions/toolkit/blob/master/docs/commands.md#problem-matchers

Added in https://github.com/python/cpython/pull/20325

Discussed with screenshots in https://discuss.python.org/t/using-github-problem-matchers-to-catch-warnings-early/4254

ezio-melotti replied:
Neat!
Thanks for sharing these links!

AA-Turner committed:
Ezio's comments

- Split to SPHINXOPTS and SPHINXERRORHANDLING
- Remove SPHINXOPTS from make check

JelleZijlstra reviewed (approved):
Looks good, but I'm no expert on CI syntax.
